src/main/java/org/opendatakit/aggregate/parser/BaseFormParserForJavaRosa.java

```diff
@@ -601,28 +579,97 @@ protected BaseFormParserForJavaRosa(String existingXml, String existingTitle, bo
     title = formTitle.replace(BasicConsts.FORWARDSLASH, BasicConsts.EMPTY_STRING);
   }
 
-  private String getNodeset(TreeElement e) {
-    String nodeset = e.getName();
-    TreeElement current = (TreeElement) e.getParent();
-    while (current != null && current.getName() != null) {
-      nodeset = current.getName() + "/" + nodeset;
-      current = (TreeElement) current.getParent();
+  private static FormDef parseDocumentIntoFormDef(Document doc) throws ODKIncompleteSubmissionData {
+    try {
+      return new XFormParser(doc).parse();
+    } catch (Exception e) {
+      throw new ODKIncompleteSubmissionData(
+          "Javarosa failed to construct a FormDef. Is this an XForm definition?", e,
+          Reason.BAD_JR_PARSE);
     }
-    return "/" + nodeset;
   }
 
-  private static List<TreeElement> flatten(TreeElement element) {
-    return childrenOf(element).stream().flatMap(e -> e.getNumChildren() == 0
-        ? Stream.of(e)
-        : childrenOf(e).stream()
-    ).collect(toList());
+  /**
+   * Reads an optional <code>odk:length</code> attribute and saves it to be used
+   * to size the corresponding storage field.
+   */
+  private void storeLengthOfBinding(Element binding) {
+    Optional
+        .ofNullable(binding.getAttributeValue(NAMESPACE_ODK, "length"))
+        .map(Integer::valueOf)
+        .ifPresent(length -> stringLengths.put(getNodeset(binding), length));
   }
 
-  private static List<TreeElement> childrenOf(TreeElement element) {
-    List<TreeElement> children = new ArrayList<>();
-    for (int i = 0, max = element.getNumChildren(); i < max; i++)
-      children.add(element.getChildAt(i));
-    return children;
+  /**
+   * Gets a copy of the given {@link Element} and saves it to be used if
+   * we need to compute the diff between the parsed xml and another xml
+   */
+  private void storeCopyOfBinding(Element binding) {
+    Element copy = new Element();
+    copy.createElement(binding.getNamespace(), binding.getName());
+    for (int i = 0; i < binding.getAttributeCount(); i++) {
+      copy.setAttribute(binding.getAttributeNamespace(i), binding.getAttributeName(i),
+          binding.getAttributeValue(i));
+    }
+    bindElements.add(copy);
+  }
+
+  private static List<Element> getBindings(Document doc) throws ODKIncompleteSubmissionData {
+    // Parse any odk:length in bindings
+    Element head = getChild(doc.getRootElement(), "head");
+    Element model = getChild(head, "model");
+    return getChildren(model, "bind");
+  }
+
+  private static Document parseXmlToDocument(String xml) throws ODKIncompleteSubmissionData {
```

Comment:
This is now virtually the same as `XFormParser.getXMLDocument` (minus some logging and an unused `CacheTable`). Could you use that one or do you find this makes the exception part clearer, perhaps?

Reply:
Thank for spotting that! I didn't realize they were so similar. Anyway, I think I'd prefer not to use `XmlFormParser.getXMLDocument`:
- It's marked as deprecated
- It does text consolidation, which we don't need in Aggregate. We save some cycles there, especially on big forms.
- It has a finer-grained exception handling, which we don't really need either. We wrap the original exception in ours, though. No information is lost.

Reply:
Wellll, it's not really marked for deprecation. The variant with just a `Reader` param is staying, it's the one with the `CacheTable` that's deprecated. But the other arguments are convincing.

Reply:
I've checked the code just in case I missed something... It's a little bit confusing:

```java
public static Document getXMLDocument(Reader reader)throws IOException{
  return getXMLDocument(reader,null);
}

@Deprecated
public static Document getXMLDocument(Reader reader,CacheTable<String> stringCache) throws IOException{
  // Stuff
}
```

It seems like both should have the `@Deprecated` annotation since the first depends on the second. Maybe we should create an issue to fix that on JR...

Reply:
The one with the reader is needed. It currently calls the more general one to avoid code duplication.